source/iterator/iterator.go

```diff
@@ -0,0 +1,322 @@
+
+package iterator
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jmoiron/sqlx"
+
+	"github.com/conduitio-labs/conduit-connector-db2/coltypes"
+	"github.com/conduitio-labs/conduit-connector-db2/source/position"
+)
+
+const (
+	trackingTablePattern = "CONDUIT_TRACKING_%s"
+
+	// tracking table columns.
+	columnOperationType = "CONDUIT_OPERATION_TYPE"
+	columnTimeCreated   = "CONDUIT_TRACKING_CREATED_DATE"
+	columnTrackingID    = "CONDUIT_TRACKING_ID"
+)
+
+// CombinedIterator combined iterator.
+type CombinedIterator struct {
+	cdc      *cdcIterator
+	snapshot *snapshotIterator
+
+	// connection string.
+	conn string
+
+	// table - table name.
+	table string
+	// trackingTable - tracking table name.
+	trackingTable string
+	// columns list of table columns for record payload
+	// if empty - will get all columns.
+	columns []string
+	// keys Names of columns what iterator use for setting key in record.
+	keys []string
+	// orderingColumn Name of column what iterator use for sorting data.
+	orderingColumn string
+	// batchSize size of batch.
+	batchSize int
+	// columnTypes column types from table.
+	columnTypes map[string]string
+}
+
+// NewCombinedIterator - create new iterator.
+func NewCombinedIterator(
+	ctx context.Context,
+	db *sqlx.DB,
+	conn, table, orderingColumn string,
+	cfgKeys, columns []string,
+	batchSize int,
+	snapshot bool,
+	sdkPosition sdk.Position,
+) (*CombinedIterator, error) {
+	var err error
+
+	it := &CombinedIterator{
+		conn:           conn,
+		table:          table,
+		columns:        columns,
+		orderingColumn: orderingColumn,
+		batchSize:      batchSize,
+		trackingTable:  fmt.Sprintf(trackingTablePattern, table),
+	}
+
+	// get column types for converting and get primary keys information
+	it.columnTypes, it.keys, err = coltypes.GetColumnTypes(ctx, db, table)
+	if err != nil {
+		return nil, fmt.Errorf("get table column types: %w", err)
+	}
+
+	it.setKeys(cfgKeys)
+
+	// create tracking table, create triggers for cdc logic.
+	err = it.SetupCDC(ctx, db)
+	if err != nil {
+		return nil, fmt.Errorf("setup cdc: %w", err)
+	}
+
+	pos, err := position.ParseSDKPosition(sdkPosition)
+	if err != nil {
+		return nil, fmt.Errorf("parse position: %w", err)
+	}
+
+	if snapshot && (pos == nil || pos.IteratorType == position.TypeSnapshot) {
+		it.snapshot, err = newSnapshotIterator(ctx, db, it.table, orderingColumn, it.keys, columns,
+			batchSize, pos, it.columnTypes)
+		if err != nil {
+			return nil, fmt.Errorf("new shapshot iterator: %w", err)
+		}
+	} else {
+		it.cdc, err = newCDCIterator(ctx, db, it.table, it.trackingTable, it.keys,
+			it.columns, it.batchSize, it.columnTypes, pos)
+		if err != nil {
+			return nil, fmt.Errorf("new shapshot iterator: %w", err)
+		}
+	}
+
+	return it, nil
+}
+
+// SetupCDC - create tracking table, add columns, add triggers, set identity column.
+func (c *CombinedIterator) SetupCDC(ctx context.Context, db *sqlx.DB) error {
```

Comment:
I believe it would be nice if all CDC related functionality would be in a single place (if possible that is).

Reply:
Will do